### DISCUSSION TITLE
Client Service Account Role Assignments Keep Getting Deleted

### DISCUSSION THREAD

**Discussion by jaymberry** [keycloak/keycloak#16810] [Q&A]
I have a Realm Created from a config on keycloak start up. It populates the database. This also creates a client with service account role assignments. This client is in the realm created at startup. 

After an undetermined amount of time, the role assignments disappear. I re-add them, and they disappear again.

I cannot figure out how this is happening. Any help would be much appreciated.
Thanks

**Comment by jaymberry** [comment]
Just a note, this is using organizations. Not sure if anyone has dealt with this.

**Comment by jaymberry** [comment]
Identified why this is happening. The service account user behind the client gets deleted and then re-created without all the original roles that we had added to it. It just gets re-created with default-roles-<realm> role.

Not sure why it gets deleted, something we need to look into. But, this user should come up with the roles established in the client. 

Any thoughts on how to make the roles of a service account user persist?

**Comment by jaymberry** [comment]
Seems to be something on our end, fixed now.

**Reply by lucianozablocki** [reply]
hey, did you find the cause of the problem? i noticed that the service account user is deleted whenever i make a PUT call to `/{realm}/clients/{id}`

**Reply by HeneryHawk** [reply]
Hi @jaymberry, @lucianozablocki
I have encountered the same problem that the service account user is deleted and recreated when the client is updated (PUT to `/{realm}/clients/{id}`). Thus the service account loses assigned roles.
Do you have a solution for this?

**Reply by jaymberry** [reply]
It was code on our end updating the user. I am no longer on the project anymore, if I recall correctly, it was code updating all users including system users which was behind the underlying service account. Developers just need to filter those users out of the provisioning. Hopefully this helps.

**Reply by HeneryHawk** [reply]
Thanks for your fast response but unfortunately this is somewhat different from our setup.
I have a small script that executes the following call.

```typescript
await adminClient.clients.update({id: client.id, realm: realm.id}, {
	redirectUris: "https://foo.bar",
	directAccessGrantsEnabled: true
});
```

If I go through it step by step using the debugger, the roles are still assigned to the service account before this line. If I execute the line, the roles are no longer assigned to the service account and the UUID of the service account user is also different. Therefore I strongly assume that it has nothing to do with our code, because nothing else is executed than just this one line of code.

**Reply by lucianozablocki** [reply]
hey, i got rid of the error by sending the entire client object when doing the PUT call

**Reply by jaymberry** [reply]
I cannot remember beyond what I mentioned above. Sorry

**Reply by lilljonas** [reply]
I faced a similar challenge where I experimented with sending different portions of the client object. Initially, I tried sending the entire object, which succeeded. Then, I gradually reduced the attributes I sent, and each attempt still worked seamlessly. However, the issue arose when I excluded the `"serviceAccountsEnabled": true` field. At that point, the process ceased to function as expected. It became apparent that this specific field must be included in the PUT request, regardless of its prior setting.

**Reply by Loki-Afro** [reply]
wanted to open an issue on that one, but i validated that against the latest version as of writing 26.2.4 and the issue appears to be resolved, i do only update the `redirectUris` via a PUT however in 23.0.7  it appears to be broken

... so i guess updating is the better option if you are able to
